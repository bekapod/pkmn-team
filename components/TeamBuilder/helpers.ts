import { anyPass, isEmpty, isNil, set, unset } from "lodash/fp";
import { ITeamMember } from "../../types";

export const validate = (
  props: {
    teamBuilderName?: string;
    teamBuilderMembers?: ITeamMember[];
  },
  state: {
    isValid: boolean;
    isTouched: boolean;
    errors: { [key: string]: string };
  },
  options: { setTouched?: boolean } = {}
) => {
  const { teamBuilderName, teamBuilderMembers } = props;
  const isInvalid = anyPass([isEmpty, isNil]);
  let updatedState = { ...state };

  if (options.setTouched) {
    updatedState = set("isTouched", true, updatedState);
  }

  updatedState =
    isInvalid(teamBuilderName) || isInvalid(teamBuilderMembers)
      ? set("isValid", false, updatedState)
      : set("isValid", true, updatedState);

  if (isInvalid(teamBuilderName)) {
    updatedState = set(
      "errors",
      { ...updatedState.errors, name: "Team name is required" },
      updatedState
    );
  } else {
    updatedState = set(
      "errors",
      unset("name", updatedState.errors),
      updatedState
    );
  }

  if (isInvalid(teamBuilderMembers)) {
    updatedState = set(
      "errors",
      { ...updatedState.errors, members: "Your team must have some pokemon" },
      updatedState
    );
  } else {
    updatedState = set(
      "errors",
      unset("members", updatedState.errors),
      updatedState
    );
  }

  return updatedState;
};
