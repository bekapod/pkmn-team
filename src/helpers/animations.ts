import { keyframes } from "styled-components/macro";

export const radialIn = `
  transform: perspective(1px) translateZ(0);
  box-shadow: 0 0 1px rgba(0, 0, 0, 0);
  position: relative;
  overflow: hidden;
  transition-property: color;
  transition-duration: 0.5s;

  &::before {
    content: "";
    position: absolute;
    z-index: -1;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    border-radius: 100%;
    transform: scale(2);
    transition-property: transform;
    transition-duration: 0.3s;
    transition-timing-function: ease-out;
  }

  &:hover::before,
  &:focus::before,
  &:active::before {
    transform: scale(0);
  }
`;

export const zoomIn = `
  transition-property: transform;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;

  &:hover,
  &:focus {
    transform: scale(1.05);
  }
`;

export const pulseFade = keyframes`
  from {
    opacity: 0.5;
    transform: scale3d(1, 1, 1);
  }

  50% {
    opacity: 1;
    transform: scale3d(1.1, 1.1, 1.1);
  }

  to {
    opacity: 0.5;
    transform: scale3d(1, 1, 1);
  }
`;

export const spin = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;
