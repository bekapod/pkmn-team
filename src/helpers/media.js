// @flow
import { css } from "styled-components/macro";

const sizes = {
  large: 1024,
  medium: 768
};

/* eslint-disable no-param-reassign */
export const media = Object.keys(sizes).reduce((accumulator, label) => {
  const size = sizes[label];
  accumulator[label] = (...args: any) => css`
    @media (min-width: ${size}px) {
      ${css(...args)};
    }
  `;
  return accumulator;
}, {});
/* eslint-enable no-param-reassign */
