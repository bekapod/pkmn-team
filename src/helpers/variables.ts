export const colors = {
  primary: "#e84118",
  primaryDark: "#c23616",
  secondary: "#273c75",
  secondaryDark: "#192a56",
  selection: "#fbc531",
  tertiary: "#fbc531",

  gray: "#7f8fa6",
  grayDark: "#353b48",
  grayDarker: "#2f3640",
  grayLight: "#dcdde1",
  white: "#f5f6fa",

  error: "#e84118",

  BUG: "#A8B820",
  DARK: "#705848",
  DRAGON: "#7038F8",
  ELECTRIC: "#F8D030",
  FAIRY: "#EE99AC",
  FIGHTING: "#C03028",
  FIRE: "#F08030",
  FLYING: "#A890F0",
  GHOST: "#705898",
  GRASS: "#78C850",
  GROUND: "#E0C068",
  ICE: "#98D8D8",
  NORMAL: "#A8A878",
  POISON: "#A040A0",
  PSYCHIC: "#F85888",
  ROCK: "#B8A038",
  SHADOW: "#604E82",
  STEEL: "#B8B8D0",
  UNKNOWN: "#68A090",
  WATER: "#6890F0"
};

export const fonts = {
  base: "'brandon-grotesque', sans-serif"
};

export const fontSizes = {
  base: 18,
  xs: 14,
  // tslint:disable-next-line:object-literal-sort-keys
  sm: 16,
  md: 22,
  lg: 32,
  xl: 48,
  xxl: 64
};

export const spacing = {
  xxl: 99,
  // tslint:disable-next-line:object-literal-sort-keys
  xl: 66,
  lg: 33,
  md: 16.5,
  sm: 8.25,
  xs: 4.125
};

export const gutters = {
  grid: 33,
  page: 40
};

export const sizes = {
  borderRadius: 11,
  borderRadiusSm: 3,
  zigzag: 33
};
