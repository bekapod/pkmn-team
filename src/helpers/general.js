// @flow
import {
  cond,
  lt,
  always,
  T,
  identity,
  propOr,
  prop,
  sort,
  equals,
  gt,
  toUpper
} from "ramda";
import uuid from "uuid/v4";
import * as variables from "./variables";
import type { Type, Pokemon } from "../types";

export const getUniqueId = uuid;

export const percentage = cond([[lt(100), always(100)], [T, identity]]);

export const getTypeColor = (type: Type) =>
  propOr(variables.colors.primary, type)(variables.colors);

export const sortTypes = sort((x: Type, y: Type) => {
  if (equals(x, y)) return 0;
  if (gt(x, y)) return 1;
  return -1;
});

export const capitalize = (word: string = "") =>
  `${toUpper(word.charAt(0))}${word.slice(1)}`;

export const capitalizePokemonName = (pokemon: Pokemon) =>
  capitalize(prop("name", pokemon));

export const formatPokemonName = (pokemon: Pokemon) =>
  `#${pokemon.pokedexId} ${capitalizePokemonName(pokemon)}`;
