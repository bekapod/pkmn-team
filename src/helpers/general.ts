import {
  always,
  cond,
  equals,
  gt,
  identity,
  lt,
  prop,
  propOr,
  sort,
  T,
  toUpper
} from "ramda";
import uuid from "uuid/v4";
import { IPokemon, Type } from "../types";
import * as variables from "./variables";

export const getUniqueId = uuid;

export const percentage = cond([[lt(100), always(100)], [T, identity]]);

export const getTypeColor = (type: Type) =>
  variables.colors[type] || variables.colors.primary;

export const sortTypes = sort((x: Type, y: Type) => {
  if (equals(x, y)) {
    return 0;
  }
  if (x > y) {
    return 1;
  }
  return -1;
});

export const capitalize = (word: string = "") =>
  `${toUpper(word.charAt(0))}${word.slice(1)}`;

export const capitalizePokemonName = (pokemon: IPokemon) =>
  capitalize(prop("name", pokemon));

export const formatPokemonName = (pokemon: IPokemon) =>
  `#${pokemon.pokedexId} ${capitalizePokemonName(pokemon)}`;
