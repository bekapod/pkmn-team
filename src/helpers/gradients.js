// @flow
import {
  compose,
  divide,
  multiply,
  identity,
  inc,
  join,
  flatten,
  append,
  prepend,
  unapply
} from "ramda";
import { percentage, getTypeColor, sortTypes } from "./general";
import type { Type } from "../types";

const getColourStopCss = (type: Type, position: number) =>
  `${getTypeColor(type)} ${position}%`;

export const getTypeGradient = (types: Array<Type>) => {
  const colourWidth = divide(100, types.length);
  const getColourStop = compose(
    percentage,
    multiply(colourWidth)
  );

  const gradientString = compose(
    join(", "),
    flatten,
    t =>
      t.map((type: Type, index: number) => [
        getColourStopCss(type, getColourStop(index)),
        getColourStopCss(type, getColourStop(inc(index)))
      ]),
    sortTypes
  )(types);

  return compose(
    join(""),
    prepend("linear-gradient(90deg, "),
    append(");"),
    unapply(identity)
  )(gradientString);
};
