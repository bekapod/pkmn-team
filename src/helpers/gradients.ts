import { compose, divide, inc, join, multiply } from "ramda";
import { Type } from "../types";
import { getTypeColor, percentage, sortTypes } from "./general";

const getColourStopCss = (type: Type, position: number) =>
  `${getTypeColor(type)} ${position}%`;

export const getTypeGradient = (types: Type[]) => {
  const colourWidth = divide(100, types.length);
  const getColourStop = compose(
    percentage,
    multiply(colourWidth)
  );

  const gradientString: string = compose(
    join(", "),
    (ts: Type[]) =>
      ts.flatMap((type: Type, index: number) => [
        getColourStopCss(type, getColourStop(index)),
        getColourStopCss(type, getColourStop(inc(index)))
      ]),
    sortTypes
  )(types);

  return `linear-gradient(90deg, ${gradientString});`;
};
