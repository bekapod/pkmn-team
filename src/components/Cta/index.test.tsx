import React from "react";
// tslint:disable-next-line:no-implicit-dependencies
import { fireEvent, render } from "react-testing-library";
import { CtaButton, CtaInternalLink } from ".";
import { renderWithRouter } from "../../helpers/testUtils";

describe("<CtaInternalLink />", () => {
  it("renders a link", () => {
    const { getByText } = renderWithRouter(
      <CtaInternalLink to="/dashboard">Dashboard</CtaInternalLink>
    );

    expect(getByText(/Dashboard/).tagName).toBe("A");
  });

  describe("when user has clicked the link", () => {
    it("directs user to correct page", () => {
      const { getByText, history } = renderWithRouter(
        <CtaInternalLink to="/dashboard">Dashboard</CtaInternalLink>
      );

      expect(
        history.entries.find(entry => entry.pathname === "/dashboard")
      ).toBeFalsy();

      fireEvent.click(getByText(/Dashboard/));

      expect(
        history.entries.find(entry => entry.pathname === "/dashboard")
      ).toBeTruthy();
    });
  });

  describe("when secondary prop is passed", () => {
    it("renders a link", () => {
      const { getByText } = renderWithRouter(
        <CtaInternalLink to="/dashboard" secondary={true}>
          Dashboard
        </CtaInternalLink>
      );

      expect(getByText(/Dashboard/).tagName).toBe("A");
    });
  });

  describe("when small prop is passed", () => {
    it("renders a link", () => {
      const { getByText } = renderWithRouter(
        <CtaInternalLink to="/dashboard" small={true}>
          Dashboard
        </CtaInternalLink>
      );

      expect(getByText(/Dashboard/).tagName).toBe("A");
    });
  });
});

describe("<CtaButton />", () => {
  it("renders a button", () => {
    const onClick = () => null;
    const { getByText } = render(
      <CtaButton onClick={onClick}>Create Team</CtaButton>
    );

    expect(getByText(/Create Team/).tagName).toBe("BUTTON");
  });

  it("calls onClick", () => {
    const spy = jest.fn();
    const { getByText } = render(
      <CtaButton onClick={spy}>Create Team</CtaButton>
    );

    expect(spy).toHaveBeenCalledTimes(0);
    fireEvent.click(getByText(/Create Team/));
    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe("when secondary prop is passed", () => {
    it("renders a button", () => {
      const onClick = () => null;
      const { getByText } = render(
        <CtaButton secondary={true} onClick={onClick}>
          Create Team
        </CtaButton>
      );

      expect(getByText(/Create Team/).tagName).toBe("BUTTON");
    });
  });

  describe("when small prop is passed", () => {
    it("renders a button", () => {
      const onClick = () => null;
      const { getByText } = render(
        <CtaButton small={true} onClick={onClick}>
          Create Team
        </CtaButton>
      );

      expect(getByText(/Create Team/).tagName).toBe("BUTTON");
    });
  });
});
