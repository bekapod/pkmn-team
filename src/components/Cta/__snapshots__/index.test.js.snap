// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CtaButton /> renders without crashing 1`] = `
<button
  className="Cta__CtaButton-sc-1qsmnol-1 cCysrS"
  onClick={[Function]}
>
  Create Team
</button>
`;

exports[`<CtaButton /> when secondary prop is passed renders without crashing 1`] = `
<button
  className="Cta__CtaButton-sc-1qsmnol-1 ktamHp"
  onClick={[Function]}
>
  Create Team
</button>
`;

exports[`<CtaButton /> when small prop is passed renders without crashing 1`] = `
<button
  className="Cta__CtaButton-sc-1qsmnol-1 dXhrin"
  onClick={[Function]}
>
  Create Team
</button>
`;

exports[`<CtaInternalLink /> renders without crashing 1`] = `
<a
  className="Cta__CtaInternalLink-sc-1qsmnol-0 isVVbW"
  href="/dashboard"
  onClick={[Function]}
>
  Dashboard
</a>
`;

exports[`<CtaInternalLink /> when secondary prop is passed renders without crashing 1`] = `
<a
  className="Cta__CtaInternalLink-sc-1qsmnol-0 kEdUQK"
  href="/dashboard"
  onClick={[Function]}
  secondary={true}
>
  Dashboard
</a>
`;

exports[`<CtaInternalLink /> when small prop is passed renders without crashing 1`] = `
<a
  className="Cta__CtaInternalLink-sc-1qsmnol-0 hEZUSK"
  href="/dashboard"
  onClick={[Function]}
  small={true}
>
  Dashboard
</a>
`;
